src/modm/platform/eth/stm32/eth.hpp

```diff
@@ -0,0 +1,696 @@
+		Pbl0                  = modm::Bit8,
+		EnhancedDescFormat    = modm::Bit7,
+		DescriptorSkipLength4 = modm::Bit6,
+		DescriptorSkipLength3 = modm::Bit5,
+		DescriptorSkipLength2 = modm::Bit4,
+		DescriptorSkipLength1 = modm::Bit3,
+		DescriptorSkipLength0 = modm::Bit2,
+		DmaArbitration        = modm::Bit1,
+		SoftwareReset         = modm::Bit0,
+	};
+	MODM_FLAGS32(DmaBusMode);
+
+	enum class
+	BurstLength : uint32_t {
+		Length1Beat   = 0b000001,
+		Length2Beats  = 0b000010,
+		Length4Beats  = 0b000100,
+		Length8Beats  = 0b001000,
+		Length16Beats = 0b010000,
+		Length32Beats = 0b100000,
+	};
+	using BurstLength_t = modm::Configuration<DmaBusMode_t, BurstLength, 0b111111, 8>;
+	using RxDmaBurstLength_t = modm::Configuration<DmaBusMode_t, BurstLength, 0b111111, 17>;
+
+	template <Peripheral peripheral, template <Peripheral _> class... Signals>
+	struct GpioConfigurator;
+	template <Peripheral peripheral, template <Peripheral _> class Signal, template <Peripheral _> class... Signals>
+	struct GpioConfigurator<peripheral, Signal, Signals...>
+	{
+		static inline void
+		configure() {
+			Signal<peripheral>::Gpio::configure(modm::platform::Gpio::OutputType::PushPull, modm::platform::Gpio::OutputSpeed::VeryHigh);
+			GpioConfigurator<peripheral, Signals...>::configure();
+		}
+	};
+	template <Peripheral peripheral>
+	struct GpioConfigurator<peripheral>
+	{
+		static inline void
+		configure() {}
+	};
+
+public:
+	template <template<Peripheral _> class... Signals>
+	static void
+	connect()
+	{
+		using Configurator = GpioConfigurator<Peripheral::Eth, Signals...>;
+		using Connector = GpioConnector<Peripheral::Eth, Signals...>;
+
+		Configurator::configure();
+		Connector::connect();
+	}
+
+	template <MediaInterface Interface>
+	static inline bool
+	initialize(uint8_t priority=5)
+	{
+		using namespace modm::literals;
+
+		Rcc::enable<Peripheral::Eth>();
+
+		NVIC_SetPriority(ETH_IRQn, priority);
+		NVIC_EnableIRQ(ETH_IRQn);
+
+		/* Select MII or RMII Mode*/
+		SYSCFG->PMC &= ~(SYSCFG_PMC_MII_RMII_SEL);
+		SYSCFG->PMC |= uint32_t(Interface);
+
+		/* Ethernet Software reset */
+		/* Set the SWR bit: resets all MAC subsystem internal registers and logic */
+		/* After reset all the registers holds their respective reset values */
+		ETH->DMABMR |= DmaBusMode_t(DmaBusMode::SoftwareReset | DmaBusMode::EnhancedDescFormat).value;
+
+		/* Wait for software reset */
+		/* Note: The SWR is not performed if the ETH_RX_CLK or the ETH_TX_CLK are
+		 * not available, please check your external PHY or the IO configuration */
+		int timeout = 1'000; // max 1ms
+		while ((DmaBusMode(ETH->DMABMR) & DmaBusMode_t(DmaBusMode::SoftwareReset)) and (timeout-- > 0)) {
+			// Wait until the PHY has reset.
+			modm::delay_us(1);
+
+			// NOTE: If the program hangs here check MII/RMII value.
+		}
+		if (not modm_assert_continue_fail_debug(timeout > 0, "eth.init",
+				"ETH::initialize() timed out on software reset!", 1))
+			return false;
```

Comment:
Related to this code...

Are there any jumpers on the Nucleo-F767ZI board that have to be changed for Ethernet to work?

Reply:
Let me check the manual...

Reply:
All jumper and solder bridges are set like described in the [Nucleo-144 manual](https://www.st.com/resource/en/user_manual/dm00244518-stm32-nucleo144-boards-mb1137-stmicroelectronics.pdf):

![grafik](https://user-images.githubusercontent.com/2820734/107638962-93fe8500-6c70-11eb-8c85-bfedad5748e0.png)

Reply:
I can't find the configuration of the pins. In my code I have this in Board::initialize():
```c++
	namespace Ethernet
	{
		using RMII_Ref_Clk = modm::platform::GpioInputA1;
		using RMII_Mdio = modm::platform::GpioA2;
		using RMII_Crs_Dv = modm::platform::GpioInputA7;
		using RMII_Tx_En = modm::platform::GpioOutputB11;
		using RMII_Tx_D0 = modm::platform::GpioOutputB12;
		using RMII_Tx_D1 = modm::platform::GpioOutputB13;
		using RMII_Mdc = modm::platform::GpioOutputC1;
		using RMII_Rx_D0 = modm::platform::GpioInputC4;
		using RMII_Rx_D1 = modm::platform::GpioInputC5;
		using Port = modm::platform::Eth<modm::Lan8720a>;
	}
```
and
```c++
		Ethernet::Port::connect<Ethernet::RMII_Ref_Clk::Refclk, Ethernet::RMII_Mdc::Mdc,
				Ethernet::RMII_Mdio::Mdio, Ethernet::RMII_Crs_Dv::Rcccrsdv,
				Ethernet::RMII_Tx_En::Txen,
				Ethernet::RMII_Tx_D0::Txd0, Ethernet::RMII_Tx_D1::Txd1,
				Ethernet::RMII_Rx_D0::Rxd0, Ethernet::RMII_Rx_D1::Rxd1>();
```

Reply:
SB13 is soldered on, but this should not cause the issue?!

Reply:
At least it starts and prints the IP:
```
Reboot: Ethernet Example
IP address: 192.168.1.1
Netmask   : 255.255.255.0
Gateway   : 0.0.0.0
DNS       : 0.0.0.0
```

Reply:
```
Reboot: Ethernet Example
IP address: 192.168.1.1
Netmask   : 255.255.255.0
Gateway   : 0.0.0.0
DNS       : 0.0.0.0
```

... and now it works!

Reply:
Ping and www are good?